_posts/24-07-2022-what_is_a_reader_and_why_it_matters.md

```diff
@@ -0,0 +1,134 @@
+
+```Scala
+def f: A => E => B = ???
+def g: B => E => C = ???
+def h: C => E => D = ???
+```
+
+There's a lot of `Reader`s just waiting in line. We can now rewrite our functions again:
+
+```Scala
+def f: A => Reader[E, B] = ???
+def g: B => Reader[E, C] = ???
+def h: C => Reader[E, D] = ???
+```
+
+### The solution
+Now we just need to compose these somehow. First of all, let's give our Reader a helper method:
+
+```Scala
+case class Reader[E, V](read: E => V) {
+  def pipeTo[W](reader: V => Reader[E, W]): Reader[E, W] = Reader { env =>
+    val result1 = read(env)
+    reader(result1).read(env)
+  }
+}
+```
+
+Then we can make ourselves a helper:
+
+```Scala
+object Reader {
+  implicit class ReaderOps[E, A, B](f1: A => Reader[E, B]) {
+    def andThenR[C](f2: B => Reader[E, C]): A => Reader[E, C] =
+      a => f1(a).pipeTo(f2)
+  }
+}
+```
+
+Et voilà, we can now write this:
+
+```Scala
+def i: A => Reader[E, D] = f andThenR g andThenR h
+```
+
+And it works!
+
+### A bit of black magic
+We can do one more trick. Let's first refactor the Reader to be able to use Scala's wonderful for-comprehensions by renaming `pipeTo` to `flatMap` and adding a `map` method:
+
+```Scala
+case class Reader[E, V](read: E => V) {
+  def map[W](f: V => W): Reader[E, W] = Reader(env => f(read(env)))
+  def flatMap[W](f: V => Reader[E, W]): Reader[E, W] = Reader { env =>
+    val result1 = read(env)
+    f(result1).read(env)
+  }
+}
+```
+
+and then define:
+
+```Scala
+def ask[E]: Reader[E, E] = Reader(identity)
+```
+
+Now we can write stuff like this:
+
+```Scala
+val readIncrementAndStringify = for {
+  intFromEnvironment <- ask[Int]
+  incremented = intFromEnvironment + 1
+} yield s"$incremented"
+```
+
+This yields a `Reader[Int, String]` value, which we can run afterwards to retrieve the final `String` when we need it - but the logic itself looks like pulling `Int`s out of thin air. Beautiful, isn't it?
+
+## The use
+One might ask: why does it matter at all? Aside of the obvious, this can serve as a dependency injection mechanism - functional and type safe. The `E` we plug into the `Reader` can be whatever we want - a default value for a given type, an input to a whole computation, a validator or a service with various utilities. We can even stuff multiple things into it!
+
+```Scala
+type Env = (UserDeserializer, UserValidator)
+
+def createUser(input: String): Reader[Env, Option[User]] = for {
+  (deserializer, validator) <- ask[Env] //assuming we use better-monadic-for plugin. Who doesn't?
+  deserialized = deserializer.deserialize(input)
+  isCorrect = validator.validate(deserialized)
+} yield if (isCorrect) Some(deserialized) else None
+```
+
+This idea is quite similar to how ZIO handles dependency injection, and how it's done in the wide functional world. Of course this implementation is quite limited - but its purpose is to give the reader (pun intended) an intuition on how `Reader`s work. Go forth and read!
```

Comment:
Add some link to ZIO dependency injection

Reply:
Added a footnote.